samples/core/callback/main.cpp

```diff
@@ -0,0 +1,341 @@
+/*
+ * Copyright (c) 2023 The Khronos Group Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// OpenCL SDK includes
+#include <CL/SDK/CLI.hpp>
+#include <CL/SDK/Context.hpp>
+#include <CL/SDK/Image.hpp>
+#include <CL/SDK/Options.hpp>
+#include <CL/Utils/Context.hpp>
+
+// standard header includes
+#include <cassert>
+#include <fstream>
+#include <future>
+#include <iostream>
+#include <map>
+#include <mutex>
+#include <tuple> // std::make_tuple
+#include <vector>
+
+// TCLAP includes
+#include <tclap/CmdLine.h>
+
+namespace {
+
+struct ReadJob
+{
+    // Although the buffer is not used in the callback, we keep a reference to
+    // the OpenCL object to keep it alive while the read is performed.
+    cl::Buffer buffer;
```

Comment:
I'm not quite following why an additional reference needs to be held here.  Shouldn't it be enough that the buffer is used by the call to `clEnqueueReadBuffer`?

Reply:
That's right. I didn't realize that `clEnqueueReadBuffer` holds ownership of the buffer. Fixed.